File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    boolean result = true;
    
    public boolean isValidBST(TreeNode root) {
        // approach: DFS
        
        helper(root, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return result;
    }
    
    private void helper(TreeNode node, double lower, double upper) {
        if (node == null) {
            return;
        }
        
        if (node.val <= lower || node.val >= upper) {
            result = false;
        }
        
        helper(node.left, lower, Math.min(upper, node.val));
        helper(node.right, Math.max(lower, node.val), upper);
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        # approach: DFS, keep track of min and max, must be within range
        
        res = True
        
        def helper(node, minimum, maximum):
            nonlocal res
            
            if not node:
                return
            
            if node.val <= minimum or node.val >= maximum:
                res = False
            
            helper(node.left, minimum, min(maximum, node.val))
            helper(node.right, max(minimum, node.val), maximum)
            
        helper(root, float('-inf'), float('inf'))
        return res
            